source/server/admin/stats_render.cc

```diff
@@ -0,0 +1,304 @@
+  addScalar(response, name, ValueUtil::numberValue(value));
+}
+
+// Buffers a JSON fragment for a text-readout stat, flushing to the response
+// buffer once we exceed JsonStatsFlushCount stats.
+void StatsJsonRender::generate(Buffer::Instance& response, const std::string& name,
+                               const std::string& value) {
+  addScalar(response, name, ValueUtil::stringValue(value));
+}
+
+// In JSON we buffer all histograms and don't write them immediately, so we
+// can, in one JSON structure, emit shared attributes of all histograms and
+// each individual histogram.
+//
+// This is counter to the goals of streaming and chunked interfaces, but
+// usually there are far fewer histograms than counters or gauges.
+//
+// We can further optimize this by streaming out the histograms object, one
+// histogram at a time, in case buffering all the histograms in Envoy
+// buffers up too much memory.
+void StatsJsonRender::generate(Buffer::Instance&, const std::string& name,
+                               const Stats::ParentHistogram& histogram) {
+  switch (histogram_buckets_mode_) {
+  case Utility::HistogramBucketsMode::NoBuckets:
+    summarizeBuckets(name, histogram);
+    break;
+  case Utility::HistogramBucketsMode::Cumulative: {
+    const Stats::HistogramStatistics& interval_statistics = histogram.intervalStatistics();
+    const UInt64Vec& interval_buckets = interval_statistics.computedBuckets();
+    const UInt64Vec& cumulative_buckets = histogram.cumulativeStatistics().computedBuckets();
+    collectBuckets(name, histogram, interval_buckets, cumulative_buckets);
+    break;
+  }
+  case Utility::HistogramBucketsMode::Disjoint: {
+    const Stats::HistogramStatistics& interval_statistics = histogram.intervalStatistics();
+    const UInt64Vec interval_buckets = interval_statistics.computeDisjointBuckets();
+    const UInt64Vec cumulative_buckets = histogram.cumulativeStatistics().computeDisjointBuckets();
+    collectBuckets(name, histogram, interval_buckets, cumulative_buckets);
+    break;
+  }
+  }
+}
+
+// Since histograms are buffered (see above), the finalize() method generates
+// all of them.
+void StatsJsonRender::finalize(Buffer::Instance& response) {
+  if (!stats_array_.empty()) {
+    flushStats(response);
+  }
+  if (!histogram_array_.empty()) {
+    ProtoMap& histograms_obj_container_fields = *histograms_obj_container_.mutable_fields();
+    if (found_used_histogram_) {
+      ASSERT(histogram_buckets_mode_ == Utility::HistogramBucketsMode::NoBuckets);
+      ProtoMap& histograms_obj_fields = *histograms_obj_.mutable_fields();
+      histograms_obj_fields["computed_quantiles"] = ValueUtil::listValue(histogram_array_);
+      histograms_obj_container_fields["histograms"] = ValueUtil::structValue(histograms_obj_);
+    } else {
+      ASSERT(histogram_buckets_mode_ != Utility::HistogramBucketsMode::NoBuckets);
+      histograms_obj_container_fields["histograms"] = ValueUtil::listValue(histogram_array_);
+    }
```

Comment:
If we always need covert `histogram_array_` to protobuf list, why not use protobuf list directly as the type of `histogram_array_`?

Reply:
Very good idea. This should improve rendering speed somewhat, though usually histograms are not the bottleneck.

I tried to do this also for stats_array_ but got test issues I didn't feel like resolving, as #20428 will make that obsolete anyway.